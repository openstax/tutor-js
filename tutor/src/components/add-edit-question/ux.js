import { action, observable, computed } from 'vendor';
import { filter, some, find } from 'lodash';

export default class AddEditQuestionUX {

  // local
  @observable showForm = false;

  // props
  @observable question;
  @observable book;
  // chapter/section ids
  @observable pageIds;

  //form
  @observable selectedChapter;
  @observable selectedChapterSection;

  constructor(props = {}) {
    this.book = props.book;
    // sections
    this.pageIds = props.pageIds;

    if(this.preSelectedChapters.length === 1) {
      this.selectedChapter = this.preSelectedChapters[0];
    }
  }

  @action.bound setShowForm(show) {
    this.showForm = show;
  }

  /**
   * Chapters that the user has selected
   */
  @computed get preSelectedChapters() {
    // return the chapters by checking the section pageIds
    return filter(this.book.chapters, bc => 
      some(bc.children, c => 
        some(this.pageIds, pi => pi === c.id)));
  }

  /**
   * Sections that the user has selected
   */
  @computed get preSelectedChapterSections() {
    if (!this.selectedChapter) {
      return null;
    }
    return filter(this.selectedChapter.children, scc => 
      some(this.pageIds, pi => pi === scc.id) && scc.isAssignable);
  }

  @action.bound setSelectedChapterByUUID(uuid) {
    this.selectedChapter = find(this.preSelectedChapters, psc => psc.uuid === uuid);
    if(this.preSelectedChapterSections.length === 1) {
      this.selectedChapterSection = this.preSelectedChapterSections[0];
    }
    else {
      this.selectedChapterSection = null;
    }
  }

  @action.bound setSelectedChapterSectionByUUID(uuid) {
    this.selectedChapterSection = find(this.preSelectedChapterSections, pscs => pscs.uuid === uuid);
  }
}
