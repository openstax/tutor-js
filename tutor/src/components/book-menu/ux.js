import { observable, action, observe, modelize } from 'shared/model'
import scrollIntoView from 'scroll-into-view';

export default class MenuUX {

    @observable currentPage;
    expanded = observable.map();

    constructor({ book, currentPage, wrapper }) {
        modelize(this);
        this.book = book;
        this.wrapper = wrapper;
        this.currentPage = currentPage;
        observe(this, 'currentPage', this.onPageChange, true);
    }

    expand(page, options = { wasSelected: true }) {
        this.eachParent(page, (node) => {
            const expandedOptions = this.expanded.get(node.pathId);
            if (!expandedOptions || !expandedOptions.wasSelected) {
                this.expanded.set(node.pathId, options);
            }
        });
    }

    collapse(page) {
        this.expanded.delete(page.pathId);
    }

    areSiblings(a, b) {
        return a.parent == b.parent;
    }

    eachParent(node, fn) {
        if (node) fn(node);
        while ((node = node.parent)) {
            fn(node);
        }
    }

    isExpanded(node) {
        return Boolean(this.expanded.get(node.pathId));
    }

    @action toggleExpansion(node, ev) {
        if (this.isExpanded(node)) {
            this.collapse(node);
        } else {
            this.expand(node);
        }
        ev.preventDefault();
    }

    wasAutoExpanded(page) {
        const options = this.expanded.get(page.pathId);
        return Boolean(options && options.wasSelected == false);
    }

    @action.bound onPageChange({ oldValue: oldPage, newValue: newPage }) {
        if (oldPage == newPage) { return; }

        // find chapters that were selected only because of the old page
        if (oldPage && !this.areSiblings(oldPage, newPage)) {
            this.eachParent(oldPage, (node) => {
                if (this.wasAutoExpanded(node)) {
                    this.collapse(node);
                }
            });
        }

        this.expand(newPage, { wasSelected: false });
        if (this.wrapper) {
            scrollIntoView(this.wrapper.querySelector(`[data-node-id="${newPage.pathId}"]`), {
                validTarget: (target) => {
                    return target == this.wrapper;
                },
            });
        }
    }

}
