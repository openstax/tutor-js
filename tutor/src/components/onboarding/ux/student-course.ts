import { computed, override, observable, action, modelize } from 'shared/model';
import { get } from 'lodash';
import UiSettings from 'shared/model/ui-settings';
import { BaseOnboarding } from './base';
import Nags from '../../../components/onboarding/nags';
import { Course, TourContext } from '../../../models';
import { Payments } from '../../../helpers/payments';
import { forceReload } from '../../../helpers/reload';

const PAY_LATER_CHOICE  = 'PL';
const TRIAL_ACKNOWLEDGED = 'FTA';

export class StudentCourseOnboarding extends BaseOnboarding {
    @observable displayPayment = false;
    @observable displayTrialActive = false;
    @observable needsTermsSigned = false

    constructor(course: Course, tourContext: TourContext) {
        super(course, tourContext);
        modelize(this);
    }

    @override get nagComponent(): any {
        if (this.needsTermsSigned) { return null; }
        const student = this.course.userStudentRecord;
        if (!student || student.is_comped) { return null; }
        if (this.displayPayment) { return Nags.makePayment; }
        if (!Payments.config.is_enabled && this.course.does_cost){
            if (!UiSettings.get(TRIAL_ACKNOWLEDGED, this.course.id)) {
                return Nags.payDisabled;
            }
        } else if (this.course.needsPayment) {
            if (student.mustPayImmediately) {
                return Nags.freeTrialEnded;
            } else if (this.displayTrialActive) {
                return Nags.freeTrialActivated;
            } else if (!UiSettings.get(PAY_LATER_CHOICE, this.course.id)) {
                return Nags.payNowOrLater;
            }
        }

        return null;
    }

    @computed get isDisplaying() {
        return Boolean(this.nagComponent);
    }

    @computed get paymentIsPastDue() {
        return get(this.course, 'userStudentRecord.mustPayImmediately', false);
    }

    @action.bound
    acknowledgeTrial() {
        UiSettings.set(TRIAL_ACKNOWLEDGED, this.course.id, true);
    }

    @action.bound
    payNow() {
        this.displayPayment = true;
    }

    @action.bound
    onAccessCourse() {
        this.displayTrialActive = false;
    }

    @action.bound
    onPayLater() {
        UiSettings.set(PAY_LATER_CHOICE, this.course.id, true);
        this.displayTrialActive = true;
        this.displayPayment = false;
    }

    @action.bound
    onPaymentComplete() {
        if (this.paymentIsPastDue) {
        // in this case we have to reload since network requests have been failing silently
            setTimeout(forceReload);
        } else {
            this.displayPayment = false;
            this.course.userStudentRecord?.markPaid();
            // start fetch tasks since they could not be fetched while student was in unpaid status
            this.course.studentTaskPlans.startFetching();
        }
    }

    @action mount() {
        super.mount();
        if (this.paymentIsPastDue) {
            this.priority = 0;
        } else {
        // make sure tasks list is up-to-date
            this.course.studentTaskPlans.refreshTasks();
        }
        this.tourContext.otherModal = this;
    }

    @action close() {
        super.close();
        this.tourContext.otherModal = undefined;
    }
}
