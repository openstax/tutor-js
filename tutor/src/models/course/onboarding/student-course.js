import {
  computed, observable, action,
} from 'mobx';
import { get } from 'lodash';
import UiSettings from 'shared/model/ui-settings';
import BaseOnboarding from './base';
import Nags from '../../../components/onboarding/nags';
import Payments from '../../payments';
import User from '../../user';

const PAY_LATER_CHOICE  = 'PL';
const TRIAL_ACKNOWLEDGED = 'FTA';
const INITIAL_LOADING_DELAY = 10;
const FETCH_INITIAL_TASKS_INTERVAL = 1000 * 60 * INITIAL_LOADING_DELAY; // every X minutes;
const REFRESH_TASKS_INTERVAL = 1000 * 60 * 60 * 4; // every 4 hours

export default class StudentCourseOnboarding extends BaseOnboarding {

  @observable displayPayment = false;
  @observable displayTrialActive = false;
  @observable refreshTasksTimer = null;

  @computed get nagComponent() {
    if (this.needsTermsSigned) { return null; }

    if (this.displayPayment) { return Nags.makePayment; }
    if (!Payments.config.is_enabled && this.course.does_cost){
      if (!UiSettings.get(TRIAL_ACKNOWLEDGED, this.course.id)) {
        return Nags.payDisabled;
      }
    } else if (this.course.needsPayment) {
      if (this.course.userStudentRecord.mustPayImmediately) {
        return Nags.freeTrialEnded;
      } else if (this.displayTrialActive) {
        return Nags.freeTrialActivated;
      } else if (!UiSettings.get(PAY_LATER_CHOICE, this.course.id)) {
        return Nags.payNowOrLater;
      }
    }

    return null;
  }

  @computed get needsTermsSigned() {
    return Boolean(User.terms_signatures_needed && !this.paymentIsPastDue);
  }

  @computed get isDisplaying() {
    return Boolean(this.nagComponent);
  }

  @computed get paymentIsPastDue() {
    return get(this.course, 'userStudentRecord.mustPayImmediately', false);
  }

  @action.bound
  acknowledgeTrial() {
    UiSettings.set(TRIAL_ACKNOWLEDGED, this.course.id, true);
  }

  @action.bound
  payNow() {
    this.displayPayment = true;
  }

  @action.bound
  onAccessCourse() {
    this.displayTrialActive = false;
  }

  @action.bound
  onPayLater() {
    UiSettings.set(PAY_LATER_CHOICE, this.course.id, true);
    this.displayTrialActive = true;
    this.displayPayment = false;
  }

  @action.bound
  onPaymentComplete() {
    this.displayPayment = false;
    this.course.userStudentRecord.markPaid();
    // fetch tasks since they could not be fetched while student was in unpaid status
    if (!this.refreshTasksTimer) {
      return this.fetchTaskPeriodically();
    }
    return Promise.resolve();
  }

  @computed get isEmptyNewStudent() {
    return Boolean(
      this.course.studentTasks.expecting_assignments_count &&
        this.course.primaryRole.joinedAgo('minutes') < 10
    );
  }

  @computed get fetchTasksInterval() {
    if (this.isEmptyNewStudent) {
      return FETCH_INITIAL_TASKS_INTERVAL - this.course.primaryRole.joinedAgo('ms');
    }
    return REFRESH_TASKS_INTERVAL;
  }

  @action.bound fetchTaskPeriodically() {
    return this.course.studentTasks.fetch().then(() => {
      this.refreshTasksTimer = setTimeout(this.fetchTaskPeriodically, this.fetchTasksInterval);
    });
  }

  mount() {
    super.mount();
    if (!this.refreshTasksTimer && !this.paymentIsPastDue) {
      this.fetchTaskPeriodically();
    }
    this.tourContext.otherModal = this;
  }

  close() {
    super.close();
    if (this.refreshTasksTimer) {
      clearInterval(this.refreshTasksTimer);
      this.refreshTasksTimer = null;
    }
    this.tourContext.otherModal = null;
  }

}
