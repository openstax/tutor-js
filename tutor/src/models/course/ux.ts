import {
    computed, observable,
} from 'mobx';


export default class CourseUX {

    static get displayCourseCost() {
        return false;
    }

    static get formattedStudentCost() {
        return '$10';
    }

  @observable course;

  constructor(course) {
      modelize(this);
      this.course = course;
  }

  @computed get dataProps() {
      return {
          'data-title':       this.course.nameCleaned,
          'data-book-title':  this.course.bookName,
          'data-appearance':  this.course.appearance_code,
          'data-is-preview':  this.course.is_preview || false,
          'data-term':        this.course.termFull,
      };
  }

  @computed get courseType() {
      return this.course.is_concept_coach ? 'cc' : 'tutor';
  }

  @computed get courseId() {
      return this.course.isNew ? 'new' : this.course.id;
  }

  get formattedStudentCost() {
      return CourseUX.formattedStudentCost;
  }

}
