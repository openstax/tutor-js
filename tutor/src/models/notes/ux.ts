import { observable, action, computed } from 'mobx';
import { readonly } from 'core-decorators';

const ERROR_DISPLAY_TIMEOUT = 1000 * 4;

class StatusMessage {

  @observable display = false;
  @observable type;
  @observable message;
  @observable hideTimer;

  @computed get icon() {
      if (this.type === 'error') { return 'exclamation-triangle'; }
      return 'info-circle';
  }

  @action.bound hide() {
      this.display = false;
      this.hideTimer = null;
  }

  show({ message, type = 'error', autoHide = false }) {
      if (this.hideTimer) {
          clearTimeout(this.hideTimer);
          this.hideTimer = null;
      }

      Object.assign(this, {
          display: true, message, type,
      });

      if (autoHide) {
          this.hideTimer = setTimeout(
              this.hide, ERROR_DISPLAY_TIMEOUT
          );
      }
  }

}

class AnnotatorUX {

  @observable isSummaryVisible = false;

  @action.bound hideSummary() {
      this.isSummaryVisible = false;
  }

  @action.bound toggleSummary() {
      this.isSummaryVisible = !this.isSummaryVisible;
  }

  @readonly statusMessage = new StatusMessage();

}

const ux = new AnnotatorUX();

export default ux;
