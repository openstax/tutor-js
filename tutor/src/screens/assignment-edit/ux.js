import { React, observable, computed, action } from 'vendor';
import Router from '../../helpers/router';
import { runInAction } from 'mobx';
import ScrollTo from '../../helpers/scroll-to';
import {
  filter, isEmpty, compact, map, get, first, difference, flatMap, omit,
} from 'lodash';
import Exercises from '../../models/exercises';
import TaskPlan, { SELECTION_COUNTS } from '../../models/task-plans/teacher/plan';
import ReferenceBook from '../../models/reference-book';
import { StepUX, Step } from './step';
import { Actions } from './actions';
import Validations from './validations';
import moment from '../../helpers/moment-range';
import Time from '../../models/time';

const TEMPLATEABLE_TYPES = ['homework', 'reading'];

export default class AssignmentUX {

  @observable isShowingSectionSelection = false;
  @observable isShowingExerciseReview = false;
  @observable isShowingPeriodTaskings;
  @observable isShowingAddTemplate = false;
  @observable exercises;
  @observable isReady = false;
  @observable sourcePlanId;
  @observable form;
  @observable activeFilter = 'all';
  @observable templates;

  constructor(attrs = null) {
    if (attrs) { this.initialize(attrs); }
  }

  @action async initialize({
    id, type, plan, course, history, due_at,
    gradingTemplates = course.gradingTemplates,
    exercises = Exercises,
    windowImpl = window,
  }) {
    if ('clone' === type) {
      if (!course.pastTaskPlans.api.hasBeenFetched) {
        await course.pastTaskPlans.fetch();
      }
      this.sourcePlanId = id;
      this.plan = course.pastTaskPlans.get(id).createClone({ course });
    } else {
      if (plan) {
        this.plan = plan;
      } else {
        this.plan = new TaskPlan({ id, course, type });
        if (id && id != 'new') {
          const existing = course.teacherTaskPlans.get(id);
          if (existing) {
            this.plan.update( existing.serialize() );
          }
        }
      }
      if (type) {
        this.plan.type = type;
      }
    }
    this.course = course;
    // don't setup steps until course and plan is set
    this.steps = new StepUX(this);
    this.actions = new Actions(this);

    if (this.plan.isNew) {
      if (this.plan.isExternal || this.plan.isEvent) {
        const opens_at = moment(Time.now).add(1, 'day').startOf('day').add(1, 'minute').toISOString();
        this.periods.map((period) =>
          this.plan.findOrCreateTaskingForPeriod(period, { opens_at }),
        );
        if (due_at) {
          this.plan.tasking_plans.forEach(tp => {
            tp.initializeWithDueAt({ dueAt: due_at, defaultOpenTime: '00:01', defaultDueTime: '21:00' });
            tp.closes_at = moment(tp.due_at).add(1, 'minute').toISOString();
          });
        }
      } else {
        this.periods.map((period) =>
          this.plan.findOrCreateTaskingForPeriod(period),
        );
      }
    } else {
      await this.plan.ensureLoaded();
    }

    if (this.canSelectTemplates) {
      // once templates is loaded, select ones of the correct type
      await gradingTemplates.ensureLoaded();
      this.templates = gradingTemplates;
      this.plan.grading_template_id = this.plan.grading_template_id || get(this.gradingTemplates, '[0].id');
    }

    this.history = history;
    this.exercises = exercises;

    await this.exercises.ensureExercisesLoaded({ course, exercise_ids: this.plan.exerciseIds });

    if (this.plan.isReading) {
      await this.referenceBook.ensureLoaded();
    }
    this.windowImpl = windowImpl;

    this.isShowingPeriodTaskings = !(this.canSelectAllSections && this.plan.areTaskingDatesSame);

    this.scroller = new ScrollTo({ windowImpl });
    this.validations = new Validations(this);
    this.isReady = true;
  }

  @computed get gradingTemplates() {
    return(filter(this.templates.array, t => t.task_plan_type == this.plan.type));
  }

  @computed get gradingTemplate() {
    return (
      this.course.gradingTemplates.array.find(t => t.id == this.form.values.grading_template_id)
    );
  }

  @computed get canSelectAllSections() {
    return Boolean(
      this.plan.isEditable && isEmpty(difference(
        map(this.course.periods.active, 'id'),
        map(this.plan.tasking_plans, 'target_id'),
      ))
    );
  }

  @computed get canSelectTemplates() {
    return TEMPLATEABLE_TYPES.includes(this.plan.type);
  }

  @computed get canInputExternalUrl() {
    return this.plan.type === 'external';
  }

  @computed get periods() {
    return filter(this.course.periods.sorted, 'isActive');
  }

  @computed get isInitializing() {
    return !this.isReady || !this.plan || this.plan.api.isPendingInitialFetch;
  }

  get referenceBook() {
    if (this._referenceBook) {
      return this._referenceBook;
    }
    if (this.plan.ecosystem_id && this.plan.ecosystem_id != this.course.ecosystem_id) {
      this._referenceBook = new ReferenceBook({ id: this.plan.ecosystem_id });
    } else {
      this._referenceBook = this.course.referenceBook;
    }
    return this._referenceBook;
  }

  @action.bound renderStep(form) {
    this.form = form;
    return <Step ux={this} />;
  }

  @action.bound navigateToStep(index) {
    this.history.push(
      Router.makePathname('editAssignment', {
        courseId: this.course.id,
        type: this.plan.type,
        id: this.plan.id || 'new',
        step: index,
      })
    );
  }

  @action.bound onComplete() {
    this.history.push(`/course/${this.course.id}`);
  }

  @action.bound onCancel() {
    this.onComplete();
  }

  get formValues() {
    return this.plan.serialize();
  }

  @computed get isApiPending() {
    return this.plan.api.isPending;
  }

  @computed get canEdit() {
    return this.plan.canEdit;
  }

  @computed get selectedPageIds() {
    return this.plan.pageIds;
  }

  @computed get selectedChapterSections() {
    return filter(map(this.selectedPages, 'chapter_section'), 'isPresent');
  }

  @computed get selectedPages() {
    return this.selectedPageIds.map(pgId => this.referenceBook.pages.byId.get(pgId));
  }

  @action.bound onSectionIdsChange(ids) {
    this.plan.settings.page_ids = ids;
  }

  @action.bound togglePeriodTasking({ target: input }) {
    const period = this.plan.course.periods.find(p => p.id == input.dataset.periodId);
    if (!period) { return; }

    const tasking = this.plan.tasking_plans.forPeriod(period);

    if (input.checked && !tasking) {
      const firstTp = first(this.plan.tasking_plans);
      const opens_at = firstTp && firstTp.opens_at;
      const due_at = firstTp && firstTp.due_at;
      this.plan.findOrCreateTaskingForPeriod(period, { opens_at, due_at });
    } else if (!input.checked && tasking) {
      this.plan.tasking_plans.remove(tasking);
    }
  }

  @action.bound togglePeriodTaskingsEnabled(ev) {
    this.isShowingPeriodTaskings = ev.target.value == 'periods';
    if (this.isShowingPeriodTaskings) {
      return;
    }
    this.periods.map(period => this.plan.findOrCreateTaskingForPeriod(period));
  }

  @action.bound onExerciseToggle(event, exercise) {
    const ex = exercise.wrapper;
    ex.isSelected = !ex.isSelected;
    ex.isSelected ? this.plan.addExercise(ex) : this.plan.removeExercise(ex);
  }

  @computed get isFetchingExercises() {
    return this.exercises.isFetching({ pageIds: this.selectedPageIds });
  }

  @computed get isShowingExercises() {
    return this._stepIndex === 2;
  }

  @action async onExercisesReview() {
    await this.exercises.ensureExercisesLoaded({
      course: this.course,
      book: this.referenceBook,
      exercise_ids: this.plan.exerciseIds,
    });
  }

  @computed get selectedExercises() {
    if (isEmpty(this.exercises)) { return []; }
    return compact(map(this.plan.exerciseIds, exId => (
      this.exercises.get(exId)
    )));
  }

  @computed get numTutorSelections() {
    return get(this.plan, 'settings.exercises_count_dynamic', 0);
  }

  @action.bound increaseTutorSelection() {
    if (this.canIncreaseTutorExercises) {
      this.plan.settings.exercises_count_dynamic = this.numTutorSelections + 1;
    }
  }

  @action.bound decreaseTutorSelection() {
    if (this.canDecreaseTutorExercises) {
      this.plan.settings.exercises_count_dynamic = this.numTutorSelections - 1;
    }
  }

  @computed get canIncreaseTutorExercises() {
    return this.canEdit && this.numTutorSelections < SELECTION_COUNTS.max;
  }

  @computed get canDecreaseTutorExercises() {
    return this.canEdit && this.numTutorSelections > SELECTION_COUNTS.min;
  }

  @computed get numExerciseSteps() {
    return flatMap(this.plan.exercises, 'content.questions').length;
  }

  @computed get totalSelections() {
    return this.numExerciseSteps + this.numTutorSelections;
  }

  isExerciseSelected(ex) {
    return this.selectedExercises.includes(ex);
  }

  @computed get displayedExercises() {
    return this.exercises.where(ex => (
      this.isExerciseSelected(ex) || (
        ex.isHomework && ex.isAssignable &&
        ex.page && this.selectedPageIds.includes(ex.page.id)
      )
    ));
  }

  @action.bound async onSaveAsDraftClick() {
    await this.saveAsDraft();
    this.onComplete();
  }

  @action.bound async saveAsDraft() {
    this.plan.update(omit(this.form.values, 'tasking_plans'));
    this.plan.is_draft = true;
    await this.saveAndCopyPlan();
  }

  @action.bound async onPublishClick() {
    this.plan.update(omit(this.form.values, 'tasking_plans'));
    this.plan.is_draft = false;
    if (!this.plan.is_published) {
      this.plan.is_publish_requested = true;
    }
    await this.saveAndCopyPlan();
    this.onComplete();
  }

  @action async saveAndCopyPlan() {
    await this.plan.save();
    await runInAction(() => {
      const destPlan = this.course.teacherTaskPlans.withPlanId(this.plan.id);
      destPlan.update(this.plan.serialize());
    });
  }

  @action.bound onChangeFilter(value) {
    this.activeFilter = value;
  }

  @action.bound onShowAddTemplate() {
    this.isShowingAddTemplate = true;
  }

  @action.bound onHideAddTemplate() {
    this.isShowingAddTemplate = false;
  }

}
