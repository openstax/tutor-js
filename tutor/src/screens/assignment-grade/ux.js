import { observable, action, computed, moment, modelize, runInAction } from 'vendor';
import { first, filter, isEmpty, findIndex } from 'lodash';
import { currentCourses, TeacherTaskStepGrade as Grade } from '../../models';
import ScrollTo from '../../helpers/scroll-to';
import UiSettings from 'shared/model/ui-settings';
import Time from 'shared/model/time';


export default class AssignmentGradingUX {

    @observable exercisesHaveBeenFetched = false;
    @observable selectedPeriod;

    @observable selectedHeadingIndex = 0;
    @observable expandGradedAnswers = false;
    @observable selectedStudentIndex = 0;
    @observable showOverview = false;

    @observable isPublishingScores = false;
    @observable selectedHeadingStudentsIsGrading = [];

    // @UiSettings.decorate('grd.hsn') hideStudentNames = false;
    // @UiSettings.decorate('grd.alpr') alphabetizeResponses = false;
    // @UiSettings.decorate('grd.soa') showOnlyAttempted = false;
    // @UiSettings.decorate('grd.sak') showAnswerKey = false;

    get hideStudentNames() { return (UiSettings.get('grd.hsn') || false) }
    set hideStudentNames(value) { return UiSettings.set('grd.hsn', value) }

    get alphabetizeResponses() { return (UiSettings.get('grd.alpr') || false) }
    set alphabetizeResponses(value) { return UiSettings.set('grd.alpr', value) }

    get showOnlyAttempted() { return (UiSettings.get('grd.soa') || false) }
    set showOnlyAttempted(value) { return UiSettings.set('grd.soa', value) }

    get showAnswerKey() { return (UiSettings.get('grd.sak') || false) }
    set showAnswerKey(value) { return UiSettings.set('grd.sak', value) }

    viewedQuestions = observable.map();
    constructor(attrs = null) {
        modelize(this);
        if (attrs) { this.initialize(attrs); }
    }

    @action async initialize({
        id, periodId, courseId, course, questionId,
        scores = course.teacherTaskPlans.withPlanId(id).scores,
        windowImpl = window,
    }) {

        this.scroller = new ScrollTo({ windowImpl });
        this.course = course || currentCourses.get(courseId);
        this.selectedPeriod = this.course.periods.active.find(p => p.id == periodId) ||
          first(this.course.periods.active);
        this.planScores = scores;

        await this.planScores.fetch();
        await this.planScores.taskPlan.fetch();
        await this.planScores.taskPlan.analytics.fetch();
        await this.planScores.ensureExercisesLoaded();

        this.exercisesHaveBeenFetched = true;

        if (questionId) {
            const index = findIndex(this.headings, (h => h.question_id == questionId ));
            if (index > -1) {
                runInAction(() => this.selectedHeadingIndex = index);
            }
        }
        else {
            let index = findIndex(this.headings, (h => !h.gradedStats.complete));
            if (index <= -1) { index = 0; }
            runInAction(() => this.selectedHeadingIndex = index);
        }
    }

    @computed get isScoresReady() { return this.planScores.api.hasBeenFetched; }
    @computed get isExercisesReady() { return this.isScoresReady && this.exercisesHaveBeenFetched; }

    @computed get scores() {
        return this.planScores.tasking_plans.find(tp => this.selectedPeriod.id == tp.period_id);
    }

    @computed get assignedPeriods() {
        return this.planScores.taskPlan.activeAssignedPeriods;
    }

    @computed get gradedResponses() {
        return filter(this.selectedHeading.studentResponses, sr => {
            if(this.showOnlyAttempted) {
                return sr.is_completed && !sr.needs_grading;
            }
            return !sr.needs_grading;
        });
    }

    @computed get needsGradingResponses() {
        return filter(this.selectedHeading.studentResponses, sr => {
            if(this.showOnlyAttempted) {
                return sr.is_completed && sr.needs_grading;
            }
            return sr.needs_grading;
        });
    }

    @computed get unAttemptedResponses() {
        return filter(this.selectedHeading.studentResponses, sr => !sr.needs_grading && !sr.is_completed);
    }

    @computed get selectedHeading() {
        return this.headings[this.selectedHeadingIndex];
    }

    @computed get isLastQuestion() {
        return this.selectedHeadingIndex === this.headings.length - 1;
    }

    @computed get headings() {
        if(!this.scores) return [];
        return this.scores.question_headings.gradable();
    }

    @computed get unGraded() {
        return filter(this.selectedHeading.studentResponses, s => s.needs_grading);
    }

    @computed get hasUnPublishedScores() {
        return this.scores.hasUnPublishedScores;
    }

    @computed get isLastStudent() {
        return this.needsGradingResponses.length === 1;
    }

    wasQuestionViewed(index) {
        return this.viewedQuestions.get(index);
    }

    markQuestionViewed(index) {
        this.viewedQuestions.set(index, true);
    }

    isResponseGraded(response) {
        return response.grader_points !== undefined;
    }

    isStudentAvailableToGrade(response) {
        return !response.student.extension || (!!response.student.extension && moment(response.student.extension.due_at).isBefore(Time.now));
    }

    @action async saveScore({ points, comment, response, doGoToOverview = false, doMoveNextQuestion = false }) {
        const grade = new Grade({ points, comment, response });
        await grade.save();
        //refetch scores after grade was saved
        await this.planScores.fetch();
        await this.planScores.taskPlan.fetch();
        await this.planScores.taskPlan.analytics.fetch();
        await this.planScores.ensureExercisesLoaded();

        if(doGoToOverview) {
            this.goToOverview();
        }
        // move to next question if any
        else if (isEmpty(this.unGraded) && doMoveNextQuestion) {
            if (this.selectedHeadingIndex < this.headings.length - 1){
                this.selectedHeadingIndex += 1;
            }
        }
        // go back to first student when grading the last student in the list and have some other students who were skipped
        else if (this.selectedStudentIndex >= this.unGraded.length) {
            this.selectedStudentIndex = 0;
        }
    }

    @computed get taskingPlan() {
        return this.planScores.taskPlan.tasking_plans.forPeriod(this.selectedPeriod);
    }

    @action.bound async onPublishScores() {
        this.isPublishingScores = true;
        await this.taskingPlan.publishScores();
        //refetch scores after publish scores was saved
        await this.planScores.fetch();
        await this.planScores.taskPlan.fetch();
        await this.planScores.taskPlan.analytics.fetch();
        await this.planScores.ensureExercisesLoaded();
        this.isPublishingScores = false;
    }

    @action.bound setSelectedPeriod(period) {
        this.selectedPeriod = period;
    }

    @action.bound goToQuestionHeading(index, expandGradedAnswers = false) {
        this.selectedHeadingIndex = index;
        this.expandGradedAnswers = expandGradedAnswers;
        this.selectedStudentIndex = 0;
        this.showOverview = false;
    }

    @action.bound goToOverview() {
        this.selectedHeadingIndex = undefined;
        this.expandGradedAnswers = false;
        this.selectedStudentIndex = 0;
        this.showOverview = true;
    }
}
