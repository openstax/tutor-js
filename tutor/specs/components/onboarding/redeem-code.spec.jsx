import { C } from '../../helpers';
import RedeemCode from '../../../src/components/onboarding/redeem-code';
import { StudentCourseOnboarding as CourseUX } from '../../../src/components/onboarding/ux/student-course'

describe('Redeem Code', () => {

    let ux;

    beforeEach(() => {
        ux = new CourseUX({});
    });

    it('renders and matches snapshot', () => {
        expect.snapshot(
            <C><RedeemCode ux={ux} /></C>
        ).toMatchSnapshot();
    });
});
