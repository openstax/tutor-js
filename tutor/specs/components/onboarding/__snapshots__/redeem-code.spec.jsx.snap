// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Redeem Code renders and matches snapshot 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  outline: none;
  max-width: 100%;
}

.c0 input {
  width: 130px;
  padding: 0 16px;
  height: 48px;
  border-width: 1px;
  border-style: solid;
}

.c0 input:first-child {
  width: 70px;
}

.c0 .has-errors input {
  background: #F8E8EA;
  border-color: #E298A0;
  color: #C22032;
}

.c0 .error-message {
  text-align: left;
  color: #C22032;
  font-size: 1.4rem;
}

.c0 h6,
.c0 p {
  margin: 0 !important;
}

.c0 .btn.btn-primary {
  max-width: 120px;
}

.c2 {
  margin: 20px 15px 50px;
}

.c3 {
  font-size: 1.4rem;
  width: 376px;
}

.c3 h6 {
  font-weight: bold;
}

.c3 p {
  min-width: 100%;
}

.c3 a {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

@media (min-width:1000px) {
  .c2 > *:not(:last-child) {
    margin-right: 16px;
  }
}

@media (max-width:600px) {
  .c2 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
  }

  .c2 .error-message {
    text-align: center;
  }

  .c2 .btn.btn-primary {
    margin-top: 1.6rem;
    max-width: 216px;
    -webkit-align-self: center;
    -ms-flex-item-align: center;
    align-self: center;
  }
}

<div
  className="openstax-debug-content"
>
  <div
    data-purpose="tour-conductor-wrapper"
  >
    <div>
      <div
        className="onboarding-nag redeem-code"
      >
        <div
          className="body"
        >
          <div
            className="c0"
          >
            <h5>
              Enter your pre-purchased access code below.
            </h5>
            <a
              aria-label="You can purchase access codes at your institution’s bookstore. Check with your bookstore for availability."
              href="#"
              onBlur={[Function]}
              onFocus={[Function]}
              onMouseOut={[Function]}
              onMouseOver={[Function]}
            >
              Where to find the access code?
            </a>
            <div
              className="c1 c2"
            >
              <div
                className=""
              >
                <input
                  autoFocus={true}
                  data-test-id="payment-code-prefix"
                  maxLength={999}
                  minLength={1}
                  onChange={[Function]}
                  placeholder="OST"
                  required={true}
                  type="input"
                />
                <span>
                   - 
                </span>
                <input
                  data-test-id="payment-code-post"
                  maxLength={10}
                  minLength={10}
                  onChange={[Function]}
                  placeholder="100234ABCD"
                  required={true}
                  type="input"
                />
                <div
                  className="error-message"
                >
                  
                </div>
              </div>
              <button
                className="btn btn-primary"
                disabled={true}
                onClick={[Function]}
                type="button"
              >
                Redeem
              </button>
            </div>
            <div
              className="c3"
            >
              <h6>
                Once redeemed, this access code is non-refundable.
              </h6>
              <p>
                If you’re unsure about enrolling in this class, we recommend taking advantage of our 
                <a
                  href="#"
                  onClick={[Function]}
                >
                  14-day free trial period
                </a>
                .
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="visible-when-debugging"
    >
      <div
        className="tour-spy-info"
      >
        available regions: []; region tour ids: []; audience tags: []; tour tags: []; eligible tours: []; TOUR RIDE: &lt;none&gt;
      </div>
    </div>
  </div>
  <a
    aria-hidden="true"
    className="debug-toggle-link"
    href="#spy"
    onClick={[Function]}
    tabIndex={-1}
  >
    π
  </a>
</div>
`;
